### ISSUE TITLE
facial landmark  regression

### ISSUE THREAD

**Issue by Rainy000** [Seanlinx/mtcnn#37]
Hello@Seanlinx,According to [issue](https://github.com/Seanlinx/mtcnn/issues/25) I finished facial landmark based on your Onet, the location results is not very robust. It looks very well on some picture but sometimes the deviation is big. How to adjust the super parameters to optimize the result?
PS: I already try some methods such as : learning rate(adjust to 0.001)/dropout(0.25)、weight decay(more bigger)..

**Comment by Seanlinx** [comment]
@Rainy000 
Try clip_gradient while initializing optimizer in train.py

**Comment by Rainy000** [comment]
@Seanlinx I will try it, thank you very much!

**Comment by Seanlinx** [comment]
@Rainy000 
Sorry, I mistook deviation for derivative, so clip gradient is not a solution.
Did you follow the ratio the author mentioned in the paper? I think raising the importance of landmark alignment task will probably help.

**Comment by Rainy000** [comment]
@Seanlinx Of course, but I just do face alignment on Onet, and the ratio of tasks is:  face detection : 1、box regression: 0.5、facial landmark: 1. The paper do face alignment in  three phases(Pnet Rnet Onet), but I do not know how to transfer the location results of Pnet and Rnet to Onet. Can you give me some suggestions?

**Comment by Seanlinx** [comment]
@Rainy000 
You can test the celebA dataset using your Rnet model and compute normalized error of predicted landmarks, then choose those with big deviations to be the training set of Onet

**Comment by Rainy000** [comment]
@Seanlinx  It is similar to the hard examples of face detection training set?

**Comment by Seanlinx** [comment]
@Rainy000 
Exactly.

**Comment by pribadihcr** [comment]
@Rainy000,

I want to fine tune the landmark also. Some times MTCNN bad in predict the landmark point in suppose 90 degree rotation face. Can you give me step by step how to do it. Thanks

**Comment by Rainy000** [comment]
@pribadihcr I have not tested the landmark result on 90 degree rotation face.  If you want to train the landmark model, you can refer to [https://github.com/Seanlinx/mtcnn/issues/25](url)

**Comment by HaoLiuHust** [comment]
@Rainy000 Can you share your implementation?

**Comment by Rainy000** [comment]
@HaoLiuHust Of course. I just implement the facial landmark based on the codes  of @Seanlinx , if you are interested in this , I will take some time to look for this project.  And how to give you?

**Comment by HaoLiuHust** [comment]
@Rainy000 Appreciate your help, you can give me a link or send to my email: haoliuhust@hotmail.com , thank you again

**Comment by Rainy000** [comment]
@HaoLiuHust  I have sent the codes package to your email , if you have some questions, we can discuss together, good luck~

**Comment by HaoLiuHust** [comment]
@Rainy000 Thank you very much, but seems I did not receive the email, could you have a look?

**Comment by HaoLiuHust** [comment]
@Rainy000 Sorry, found it

**Comment by xiaoxiongli** [comment]
Dear  Rainy000,
       @Rainy000 
       
Could you please send me your code,  i train a model base on Seanlinx's code, And i also need add face landmark detection. Please send me and let's discuss it.   my email:  19295656@qq.com

thank you very much! ^_^

**Comment by Rainy000** [comment]
@xiaoxiongli Sorry, just see  your message. My company does not have network, I will send you when I go home~

**Comment by xiaoxiongli** [comment]
@Rainy000  it is very kind of you～ thank you

**Comment by linsonwang** [comment]
hi  @Rainy000, could you please send me a copy of face landmark detection? Many thanks!

**Comment by Rainy000** [comment]
@linsonwang  give me your email~

**Comment by Mumu16** [comment]
@Rainy000 hi Rainy000, could you please send me a copy of facial landmark detection? My email: jing7994@163.com. Thanks in advance!

**Comment by zzzzzz0407** [comment]
As your code is popular, i think you can share you code on the github.

**Comment by KiXiang** [comment]
@Rainy000 hi Rainy000, could you please send me a copy of facial landmark detection? My email:809209272@qq.com

**Comment by Rainy000** [comment]
@KiXiang @Mumu16  OK, GOOD LUCK~

**Comment by Rainy000** [comment]
@zHanami  Good ideas, but I am not familiar with how to upload code. The most important key is that the code is implemented based on @Seanlinx , so it's not good if I upload to my account.

**Comment by mrgjbd** [comment]
@Rainy000 hello Rainy000,I get stuck on Landmarks, could you please send me a copy of facial landmark detection?  Thanks a lot，email：494587502@qq.com

**Comment by OceanWong1991** [comment]
@Rainy000 Hi Rainy000! recently, I rewrite MTCNN by TensorFlow, but the Landmark predict is not precise, can you share your code with me, Thanks very much! email : OceanWong1990@gmail.com

**Comment by Rainy000** [comment]
@OceanWong1991  Sorry, I quit from last company and have not backup copy of this project. You can ask for help to upstairs user, or implement it refer to issue #25.

**Comment by OceanWong1991** [comment]
Thx! Anyway 

    
        
            
                
                    
                        
                    
                    
                        
                            王学忠
                            
                            
                                
                                    邮箱：oceanwong1991@gmail.com
                                
                        
                    
                
            
        
    
签名由 网易邮箱大师 定制
On 05/31/2018 11:29, freecloud wrote: @OceanWong1991  Sorry, I quit from last company and have not backup copy of this project. You can ask for help to upstairs user, or implement it refer to issue #25.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"@context":"http://schema.org","@type":"EmailMessage","potentialAction":{"@type":"ViewAction","target":"https://github.com/Seanlinx/mtcnn/issues/37#issuecomment-393390319","url":"https://github.com/Seanlinx/mtcnn/issues/37#issuecomment-393390319","name":"View Issue"},"description":"View this Issue on GitHub","publisher":{"@type":"Organization","name":"GitHub","url":"https://github.com"}}
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/Seanlinx/mtcnn","title":"Seanlinx/mtcnn","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/Seanlinx/mtcnn"}},"updates":{"snippets":[{"icon":"PERSON","message":"@Rainy000 in #37: @OceanWong1991  Sorry, I quit from last company and have not backup copy of this project. You can ask for help to upstairs user, or implement it refer to issue #25."}],"action":{"name":"View Issue","url":"https://github.com/Seanlinx/mtcnn/issues/37#issuecomment-393390319"}}}
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "37567f93-e2a7-4e2a-ad37-a9160fc62647",
"title": "Re: [Seanlinx/mtcnn] facial landmark  regression (#37)",
"sections": [
{
"text": "",
"activityTitle": "**freecloud**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@Rainy000",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"Seanlinx/mtcnn\",\n\"issueId\": 37,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close issue",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueClose\",\n\"repositoryFullName\": \"Seanlinx/mtcnn\",\n\"issueId\": 37\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/Seanlinx/mtcnn/issues/37#issuecomment-393390319"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 226978497\n}"
}
],
"themeColor": "26292E"
}

**Comment by tzhang2014** [comment]
@Rainy000 can you share your code for training landmark with me ? thanks ! email :416734877@qq.com

**Comment by fdong0598** [comment]
@Rainy000 can you share your code for training landmark with me ? thanks ! email :1158563915@qq.com

**Comment by githublet** [comment]
@Rainy000 前辈，你好，对于landmark的训练我尝试了较长的时间，同你一样，一些还可以，一些效果很差，可以指点一下嘛？，tay train landmark long time,result same with you with some picture will but some big deviation，email :kylewo@foxmail.com

**Comment by Rainy000** [comment]
@githublet Try to expand your trainset, and keep the OHEM  turning on. Maybe, you can try some other percentage.

**Comment by githublet** [comment]
@Rainy000 
my train list like below : 
landmark: -1 -1 -1 -1  -1 l1 l2 l3 l4 l5 l6 l7 l8 l9
and I also try landmark label like below:
landmark: -1 p0 p1 p2 p3 l0 l1 l2 l3 l4 l5 l6 l7 l8 l9
witch did work?
also the doubt about neg label：
 0 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
or all zeros?
you mean landmark with OHEM ?

**Comment by Rainy000** [comment]
@githublet  If you train face detection and landmark jointly, the first label ( -1 ) represent part example which with IOU between [0.45,0.6], and used to  box regression.  you can choose other value ( such as -2 ) represent landmark examples . So the train list could be below:
img_path -2 x1 y1 x2 y2 x3 y3 x4 y4 x5 y5
Of course, the landmark label used to jointly training could be below
img_path -2 0 0 0 0 x1 y1 x2 y2 x3 y3 x4 y4 x5 y5        and the value of x,y belong to (0,1)
landmark training is regression problem, it hasn't neg example. OHEM trick can impose the robustness in complex scene. BTW, witch->which?

**Comment by Rainy000** [comment]
@githublet   From scratch.  I know -1 mean ignore label, but you have to  know which img is landmark sample. When training face detection, you can ignore part samples and landmark samples with label -1, but when training landmark ,you have to pick out the landmark samples with a distinguished label. 
I am chinese of course.

**Comment by githublet** [comment]
@Rainy000  
好吧，那好吧。。。
landmark 不是一起训练的嘛?为嘛要区别对待？像congweilin那样?

**Comment by Rainy000** [comment]
@githublet The reason why give landmark samples a distinguished label is to training  cls, box regression, landmark regression together. You should look this project more carefully.

**Comment by githublet** [comment]
@Rainy000 
I use reg[0,:] equal 4 or 10 to distinguish part or landmark ,some people tell me value in the  interval [0,1]is hard to train

**Comment by PhenomenalOnee** [comment]
@Rainy000  
can you provide your code and data for training the three nets seperately for both face landmark and detection

**Comment by cuiyong127** [comment]
@Rainy000 hello Rainy000,I get stuck on Landmarks, can you share your code for training landmark with me ? thanks ! email :359894979@qq.com

**Comment by gm19900510** [comment]
hi Rainy000, could you please send me a copy of facial landmark detection? My email:1025304567@qq.com

**Comment by cuiyong127** [comment]
hi,i also need the code of facial landmark detection.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"gm19900510"<notifications@github.com&gt;;
发送时间:&nbsp;2020年1月2日(星期四) 下午3:28
收件人:&nbsp;"Seanlinx/mtcnn"<mtcnn@noreply.github.com&gt;;
抄送:&nbsp;"Richar"<359894979@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [Seanlinx/mtcnn] facial landmark  regression (#37)




hi Rainy000, could you please send me a copy of facial landmark detection? My email:1025304567@qq.com

**Comment by gm19900510** [comment]
请更新代码，模型进行替换




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"崔勇"<notifications@github.com&gt;;
发送时间:&nbsp;2020年1月16日(星期四) 下午4:50
收件人:&nbsp;"Seanlinx/mtcnn"<mtcnn@noreply.github.com&gt;;
抄送:&nbsp;"1025304567"<1025304567@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [Seanlinx/mtcnn] facial landmark  regression (#37)



hi,i also need the code of facial landmark detection. 
 
 
 ------------------&amp;nbsp;原始邮件&amp;nbsp;------------------ 
 发件人:&amp;nbsp;"gm19900510"<notifications@github.com&amp;gt;; 
 发送时间:&amp;nbsp;2020年1月2日(星期四) 下午3:28 
 收件人:&amp;nbsp;"Seanlinx/mtcnn"<mtcnn@noreply.github.com&amp;gt;; 
 抄送:&amp;nbsp;"Richar"<359894979@qq.com&amp;gt;;"Comment"<comment@noreply.github.com&amp;gt;; 
 主题:&amp;nbsp;Re: [Seanlinx/mtcnn] facial landmark  regression (#37) 
 
 
 
 
 hi Rainy000, could you please send me a copy of facial landmark detection? My email:1025304567@qq.com
